src/flint/test/test.py

```diff
@@ -1813,6 +1830,324 @@ def test_fmpz_mod_dlog():
             x = g.discrete_log(a)
             assert g**x == a
 
+def test_fmpz_mod_poly():
+    from flint import fmpz_poly, fmpz_mod_poly, fmpz_mod_poly_ctx, fmpz_mod_ctx, fmpz
+
+    # fmpz_mod_poly_ctx tests
+    F = fmpz_mod_ctx(11)
+    R1 = fmpz_mod_poly_ctx(F)
+    R2 = fmpz_mod_poly_ctx(11)
+    R3 = fmpz_mod_poly_ctx(13)
+
+    assert raises(lambda: fmpz_mod_ctx("AAA"), TypeError)
+    assert raises(lambda: fmpz_mod_ctx(-1), ValueError)
+    assert (R1 == R1) is True
+    assert (R1 == R2) is True
+    assert (R1 != R3) is True
+    assert (R1 != "AAA") is True
+
+    assert (hash(R1) == hash(R1)) is True
+    assert (hash(R1) == hash(R2)) is True
+    assert (hash(R1) != hash(R3)) is True
+
+    assert str(R1) == "Context for fmpz_mod_poly with modulus: 11"
+    assert str(R1) == str(R2)
+    assert repr(R3) == "fmpz_mod_poly_ctx(13)"
+
+    assert R1.modulus() == 11
+    
+    assert R1.is_prime()
+    assert R1.zero() == 0
+    assert R1.one() == 1
+    assert R1.gen() == R1([0,1])
+
+    # Random testing
+    f = R1.random_element()
+    assert f.degree() == 3
+    f = R1.random_element(degree=5, monic=True)
+    assert f.degree() == 5
+    assert f.is_monic()
+    f = R1.random_element(degree=100, irreducible=True)
+    assert f.degree() == 100
+    assert f.is_irreducible()
+    f = R1.random_element(degree=1, monic=True, irreducible=True)
+    assert f.degree() == 1
+    assert f.is_irreducible()
+    assert f.is_monic()
+    assert raises(lambda: R1.random_element(degree=-123), ValueError)
+    assert raises(lambda: R1.random_element(monic="A"), ValueError)
+    assert raises(lambda: R1.random_element(irreducible="A"), ValueError)
+
+
+    # Conversion tests
+    F = fmpz_mod_ctx(11)
+    F_other = fmpz_mod_ctx(10)
+    R = fmpz_mod_poly_ctx(F)
+    R_other = fmpz_mod_poly_ctx(F_other)
+
+    assert raises(lambda: fmpz_mod_poly(1, "A"), TypeError) # Need a valid context
+    assert raises(lambda: R(R_other([1,2,3])), ValueError), f"{R(R_other([1,2,3])) = }" # moduli must match
+    assert raises(lambda: R(F_other(2)), ValueError) # moduli must match
+    assert raises(lambda: R([F(1), F_other(2)]), ValueError) # moduli must match
+    assert raises(lambda: R([F(1), "A"]), TypeError) # need to be able to cast to fmpz_mod
+
+    f1 = R([int(-1),int(-2),int(-3)])
+    f2 = R([fmpz(-1),fmpz(-2),fmpz(-3)])
+    f3 = R([F(-1),F(-2),F(-3)])
+    f4 = R(fmpz_poly([-1, -2, -3]))
+    f5 = R(f4)
+
+    assert str(f1) == "8*x^2 + 9*x + 10"
+    assert str(f2) == "8*x^2 + 9*x + 10"
+    assert str(f3) == "8*x^2 + 9*x + 10"
+    assert str(f4) == "8*x^2 + 9*x + 10"
+    assert str(f5) == "8*x^2 + 9*x + 10"
+
+    f1 = R(5)
+    f2 = R(fmpz(6))
+    f3 = R(F(7))
+    assert str(f1) == "5"
+    assert str(f2) == "6"
+    assert str(f3) == "7"
+
+    # Printing
+    f = R([5, 6, 7, 8])
+    assert str(f) == "8*x^3 + 7*x^2 + 6*x + 5"
+    # assert repr(f) == "fmpz_mod_poly([5, 6, 7, 8], fmpz_mod_poly_ctx(11))"
```

Comment:
I guess we need to decide what this is supposed to do. I would also include a `repr(f)` test for the current behaviour along with a comment indicating that maybe the behaviour should be changed.

Reply:
Yeah I'm not really happy with even the `repr(fmpz_mod(11, ctx))` output at the moment, I think it's a bit bulky and just outputting `11` would be nicer on the eyes? I'm not sure what to do.

Reply:
There is a bit of inconsistency:
```python
In [20]: f(2)
Out[20]: fmpz_mod(17, 163)

In [21]: f
Out[21]: 3*x^2 + 2*x + 1
```

Reply:
Personally I think that `repr` should show what representation you have like:
```python
In [28]: print(str(gmpy2.mpz(2)))
2

In [29]: print(repr(gmpy2.mpz(2)))
mpz(2)
```
Basically that is the purpose of `repr` in my mind. When you want a nicer dispaly you should use print/str.  The awkward thing is that some builtin types like `list` will always use `__repr__` recursively:
```python
In [30]: print([gmpy2.mpz(2)])
[mpz(2)]
```
One way to control this is just to have `repr` always show the pretty form but then that defeats the point of `repr` in general. Another way is to have printer control for example the context settings could be used. Another way is just to take control of printing in the console (this is what SymPy and Sage do).

I think it is something that should be decided on in general for python-flint and we should have a separate issue to discuss it.

For now I would just make sure that there is a test covering repr with its current behaviour along with a comment noting that the tested output is potentially not ideal.

Reply:
Yeah at the moment for the polynomial `repr` is the same as `str` but for `fmpz_mod` we have a more verbose `repr` compared to the `str` method. I think I prefer what `str` does.

Reply:
> I think I prefer what `str` does.

What `str` does is nice for the situation where it is intended: pretty printing.

What `repr` should do is intended for a different situation. It is very often useful to know that you have an `fmpz` rather than an `int` or that your `x^2 - 1` is in `F(163)[x]` rather than `Z[x]`, and is a python-flint type rather than a Sage/SymPy type etc.